docs/get-details/dapps/smart-contracts/guidelines.md

```diff
@@ -0,0 +1,374 @@
+* Check as many transaction fields (accessible via `txn`/`Txn` in TEAL/PyTEAL) as possible to constrain the transactions approved by the smart signature as much as possible to avoid security issues. In particular (this is not an exhaustive list):
+    * Always check the `Type` or `TypeEnum` field to ensure the transaction is of the expected type.
+    * Always check the following fields: `RekeyTo`, `CloseRemainderTo` (for payments), and `AssetCloseTo` (for asset transfers).
+        * For most cases, these fields must be set to the `ZeroAddress`.
+    * Follow the guidelines for transaction validation and for fee management above.
+* Be careful that the same smart signature can be used for multiple transactions in a group:
+    * Either check the current transaction fields using `txn` / `Txn.` (instead of `gtxn` / `Gtxn`);
+    * Or use `gtxn` / `Gtxn` for an index `i` and verify that the current transaction is the `i`-th ().
+    * Not doing the above can allow an attacker to call twice the smart signature in the same group: the first transaction obeys the rule of the smart signature and is safe, but the second one is unconstrained and can allow the attacker to siphon out all funds.
+* If the use case permits it, set up an expiration mechanism:
+    * Contract accounts should allow the sender to reclaim their assets after a certain round/time (`txn FirstValid > NNNN` or `txn FirstValidTime > TTTT`).
+    * Delegated logic signatures should specify an expiration round in the future after which they will no longer function.
+* Be extra careful when combining multiple contract accounts in a group of transaction and to perform multiple checks. The simplest solution is to enforce the transaction group size 
+* Remember that, as for smart contracts, the smart signature code, the transaction fields, and the arguments of the smart signature are all **public**. In particular:
+    * An attacker can replay a transaction signed by a smart signature (whether in contract account or logic signature mode).
+    * For example, a delegated logic signature allowing approving any change offline transaction is insecure. The first time it is used, an attacker can replay it many times and empty the account (using all the fees)! The delegated logic signature needs to 
+* Be careful about the fact that, contrary to arguments of smart contracts, arguments of smart signatures are **NOT** signed by the sender account and are **NOT** part of the computation of the group ID. In other words, an attacker can intercept a transaction signed by a smart signature and change its arguments (as long as it is still accepted by the smart signature).
+  
+
+## Additional resources
+
+* [Trails of Bits guideslines](https://github.com/crytic/building-secure-contracts/tree/master/not-so-smart-contracts/algorand)
+* [ASC Security Guidelines from Joe Polny](https://github.com/joe-p/algo-edu/blob/master/resources/en-US/asc_security.md) on which this document is partially based
+
+
```

Comment:
One extra guideline we need to discuss is min balance for boxes. If a smart contract is deleted that has existing boxes, those algos are locked forever. The guideline should say delete all boxes and recover min balance before deleting a contract. Also in the beginning when talking about escrow or account contracts, it may be good to point out that both smart sigs and smart contracts can be escrows. This will further drive your point to just use smart contracts.

Reply:
I've added boxes there:
 https://github.com/algorand/docs/pull/978/files#diff-60f54805ec9cef45e078c80bd5ecf80c5db5182486db5d18c645412492d900bcR206

and comment about escrow there:
https://github.com/algorand/docs/pull/978/files#diff-60f54805ec9cef45e078c80bd5ecf80c5db5182486db5d18c645412492d900bcR331